File: src/main/java/top/yunp/asws/engine/JsRuntime.java
/*
@author https://yunp.top
 */
package top.yunp.asws.engine;

import jakarta.servlet.http.HttpServletResponse;
import org.graalvm.polyglot.Context;
import org.graalvm.polyglot.Source;
import top.yunp.asws.http.IHttpRequest;
import top.yunp.asws.http.IHttpResponse;

import java.io.IOException;
import java.util.Queue;
import java.util.concurrent.ConcurrentLinkedQueue;

public class JsRuntime {

    private final Source entrypoint;
    private final String mainClassName;
    private final Queue<Context> contexts = new ConcurrentLinkedQueue<>();

    public JsRuntime(Source entrypoint) {
        this(entrypoint, null);
    }

    public JsRuntime(Source entrypoint, String mainClassName) {
        this.entrypoint = entrypoint;
        this.mainClassName = mainClassName;
    }

    private Context accquireContext() throws IOException {
        Context context = contexts.poll();
        if (context == null) {
            context = Context.newBuilder(Languages.JS)
                    .allowAllAccess(true)
                    .option("js.commonjs-require", "true")
                    .build();

            context.eval(entrypoint);

            if (mainClassName != null) {
                context.eval(Languages.JS, "new " + mainClassName + "()");
            }
        }
        return context;
    }

    private void recycleContext(Context context) {
        contexts.offer(context);
    }

    public void handle(IHttpRequest req, IHttpResponse res) throws IOException {
        var context = accquireContext();
        try {
            var func = context.eval("js", "application");
            func.executeVoid(req, res);
        } catch (Throwable e) {
            res.setStatus(HttpServletResponse.SC_INTERNAL_SERVER_ERROR);
            res.write("Internal server error");
            e.printStackTrace();
        }
        recycleContext(context);
    }
}

File: src/main/kotlin/top/yunp/aog/engine/JsRuntime.kt
/*
Created on 2025/11/15
@author https://yunp.top
 */
package top.yunp.aog.engine

import io.ktor.http.*
import io.ktor.server.response.*
import io.ktor.server.routing.*
import io.ktor.server.websocket.*
import kotlinx.coroutines.future.asCompletableFuture
import kotlinx.coroutines.launch
import org.graalvm.polyglot.Context
import org.graalvm.polyglot.Source
import top.yunp.aog.async.IOScope
import java.util.*
import java.util.concurrent.ConcurrentLinkedQueue


class JsRuntime(private val source: Source, private val mainClass: String? = null) {

    private val contexts: Queue<Context?> = ConcurrentLinkedQueue<Context?>()

    private fun acquireContext(): Context {
        var context: Context? = contexts.poll()
        if (context == null) {
            context = Context.newBuilder(Languages.JS)
                .allowAllAccess(true)
                .option("js.commonjs-require", "true")
                .build()

            context.eval(source)

            if (mainClass != null) {
                context.eval(Languages.JS, "new $mainClass()")
            }
        }
        return context
    }

    private fun recycleContext(context: Context?) {
        contexts.offer(context)
    }

    fun handle(routingContext: RoutingContext) {
        val context: Context = acquireContext()
        try {
            val func = context.eval("js", "application")
            func.executeVoid(JsHttpContext(routingContext = routingContext))
        } catch (e: Throwable) {
            IOScope.launch {
                routingContext.call.respondText("Internal server error", status = HttpStatusCode.InternalServerError)
            }.asCompletableFuture().get()
            e.printStackTrace()
        }
        recycleContext(context)
    }

    fun handle(session: DefaultWebSocketServerSession) {
        val context: Context = acquireContext()
        try {
            val func = context.eval("js", "application")
            func.executeVoid(JsHttpContext(webSocketSession = session))
        } catch (e: Throwable) {
            e.printStackTrace()
        }
        recycleContext(context)
    }
}